multi/nrf91-multi/uart.c

```diff
@@ -0,0 +1,537 @@
+ * Phoenix-RTOS
+ *
+ * nRF91 UART driver
+ *
+ * Copyright 2023, 2024 Phoenix Systems
+ * Author: Damian Loewnau
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+
+#include <errno.h>
+#include <stdint.h>
+#include <stdlib.h>
+#include <stdbool.h>
+#include <string.h>
+#include <paths.h>
+#include <posix/utils.h>
+#include <sys/msg.h>
+#include <sys/pwman.h>
+#include <sys/interrupt.h>
+#include <sys/file.h>
+#include <sys/threads.h>
+#include <sys/types.h>
+#include <libtty.h>
+#include <libtty-lf-fifo.h>
+
+#include <board_config.h>
+
+#include "nrf91-multi.h"
+#include "common.h"
+#include "uart.h"
+
+#define UART_CNT (UART0 + UART1 + UART2 + UART3)
+
+#define THREADS_CNT     3
+#define THREAD_STACKSZ  768
+#define THREAD_PRIORITY 1
+
+typedef struct {
+	char stack[512] __attribute__((aligned(8)));
+
+	volatile unsigned int *base;
+	volatile char *txDma;
+	volatile char *rxDma;
+
+	unsigned char stopBits;
+	int parity;
+	int baud;
+	int cnt;
+	bool init;
+	bool resetCnt;
+
+	handle_t cond;
+	handle_t irqlock;
+
+	libtty_common_t ttyCommon;
+
+	volatile bool rxdrdy;
+	int rxdCount;
+} uart_ctx_t;
+
+
+static struct {
+	unsigned char stack[THREADS_CNT][THREAD_STACKSZ] __attribute__((aligned(8)));
+	uart_ctx_t ctx[UART_CNT];
+
+	unsigned int port;
+} uart_common;
+
+
+enum { uarte_startrx = 0,
+	uarte_stoprx,
+	uarte_starttx,
+	uarte_stoptx,
+	uarte_flushrx = 11,
+	uarte_events_cts = 64,
+	uarte_events_ncts,
+	uarte_events_rxdrdy,
+	uarte_events_endrx = 68,
+	uarte_events_txdrdy = 71,
+	uarte_events_endtx,
+	uarte_events_error,
+	uarte_events_rxto = 81,
+	uarte_events_rxstarted = 83,
+	uarte_events_txstarted,
+	uarte_events_txstopped = 86,
+	uarte_inten = 192,
+	uarte_intenset,
+	uarte_intenclr,
+	uarte_errorsrc = 288,
+	uarte_enable = 320,
+	uarte_psel_rts = 322,
+	uarte_psel_txd,
+	uarte_psel_cts,
+	uarte_psel_rxd,
+	uarte_baudrate = 329,
+	uarte_rxd_ptr = 333,
+	uarte_rxd_maxcnt,
+	uarte_rxd_amount,
+	uarte_txd_ptr = 337,
+	uarte_txd_maxcnt,
+	uarte_txd_amount,
+	uarte_config = 347 };
+
+
+enum { no_parity,
+	parity_even
+};
+
+
+enum { baud_9600 = 0x00275000,
+	baud_115200 = 0x01d60000,
+	baud_230400 = 0x03b00000,
+	baud_250000 = 0x04000000,
+	baud_460800 = 0x07400000,
+	baud_921600 = 0x0f000000,
+	baud_1000000 = 0x10000000 };
+
+
+enum { uart0 = 0,
+	uart1,
+	uart2,
+	uart3 };
+
+
+/* indicates whether the tx line is ready to transmit new data */
+static bool uart_txready(uart_ctx_t *ctx)
+{
+	bool ret;
+	ret = ctx->init;
+
+	mutexLock(ctx->irqlock);
+	/* before the first transaction tx is also ready even if the event hasn't occurred */
+	if (ctx->init == false) {
+		/* true means that the data has been sent from txd */
+		ret = ctx->base[uarte_events_txdrdy];
+		ctx->base[uarte_events_txdrdy] = 0u;
+		ctx->init = false;
+	}
+	mutexUnlock(ctx->irqlock);
+
+	return ret;
+}
+
+
+static int uart_irqHandler(unsigned int n, void *arg)
+{
+	uart_ctx_t *ctx = (uart_ctx_t *)arg;
+
+	if (ctx->base[uarte_events_rxdrdy] != 0u) {
```

Comment:
Hmm, it was done, because of MISRA rule 18.4. I can change it back, but then it would be a good idea to mention such recommendation in our coding convention.

Reply:
Good point. Perhaps we should change our convention then? Something to think about. I personally like `[]` style

Reply:
Yeah, let's discuss it during some call/meeting.

Reply:
Ok, so before the new establishments I'd like to add this in the form compliant with the rest of our code for now.